net_util/src/queue_pair.rs

```diff
@@ -288,6 +292,51 @@ impl RxVirtio {
     }
 }
 
+#[derive(Default, Clone)]
+struct IovecBuffer(Vec<libc::iovec>);
+
+// SAFETY: Implementing Send for IovecBuffer is safe as the pointer inside is iovec.
+// The iovecs are usually constructed from virtio descriptors, which are safe to send across
+// threads.
+unsafe impl Send for IovecBuffer {}
+// SAFETY: Implementing Sync for IovecBuffer is safe as the pointer inside is iovec.
+// The iovecs are usually constructed from virtio descriptors, which are safe to access from
+// multiple threads.
+unsafe impl Sync for IovecBuffer {}
+
+impl IovecBuffer {
+    const fn new() -> Self {
+        IovecBuffer(Vec::new())
```

Comment:
Might as well do `Vec::with_capacity(4)` as from experimentation that is a sensible starting number to use (and avoid the initial reallocations.)

I'm still not 100% convinced this is better than smallvec in terms of maintenance overhead (avoiding heap allocations like this is exactly why we started using smallvec)

But I can see you've put a lot of effort into this work - so with the change I propose above lgtm.

Reply:
Sure. I've updated it as 4 as you suggested.

> But I can see you've put a lot of effort into this work - so with the change I propose above lgtm.

Please don't do it just because this reason. I believe I can convince you of my point of view. This is a small change and didn't cost much time, but in order to justify this change I read some kernel code and did benchmarks, which was very rewarding and I think it is good.

I did a simple benchmark(but only simulated not with vmm) here: https://gist.github.com/ihciah/dcfaf13393bfada1ff8f372b4fa16e3f

And the result shows even for len=3, borrowed vec only cost 48% CPU time of smallvec(N=4). I believe it is because some stack copying cost is saved. For longer data, borrowed vec saves more. In the link above there's more results and reproducing guide.